cloudinit/sources/DataSourceWSL.py

```diff
@@ -0,0 +1,234 @@
+    """
+    from cloudinit.settings import MACHINE_ID_FILE
+
+    if util.wait_for_files([MACHINE_ID_FILE], 2.0):
+        LOG.debug("%s file not found", MACHINE_ID_FILE)
+        return None
+
+    return util.load_file(MACHINE_ID_FILE, decode=True)
+
+
+def candidate_user_data_file_names(instance_name) -> List[str]:
+    """
+    Return a list of candidate file names that may contain user-data
+    in some supported format, ordered by precedence.
+    """
+    distribution_id, _, release_codename = util.get_linux_distro()
+
+    return [
+        # WSL instance specific:
+        "%s.user-data" % instance_name,
+        # release codename specific
+        "%s-%s.user-data" % (distribution_id, release_codename),
+        # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
+        "%s-all.user-data" % distribution_id,
+        # generic, valid for all WSL distros and instances.
+        "config.user-data",
+    ]
+
+
+class DataSourceWSL(sources.DataSource):
+    dsname = "WSL"
+
+    def __init__(self, sys_cfg, distro, paths):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths)
+        self._network_config = sources.UNSET
+        self.dsmode = sources.DSMODE_LOCAL
+        self.distro = distro
+        self.instance_name = instance_name()
+
+    def find_user_data_file(self) -> Optional[PurePath]:
+        """
+        Finds the most precendent of the candidate files that may contain
+        user-data, if any, or None otherwise.
+        """
+        profile_dir = win_user_profile_dir()
+        seed_dir = os.path.join(profile_dir, ".cloud-init")
+        if not os.path.isdir(seed_dir):
+            LOG.error("%s directory doesn't exist.", seed_dir)
+            return None
+
+        # Notice that by default file name casing is irrelevant here. Windows
+        # filenames are case insensitive. Even though accessed through Linux,
+        # path translation just works with whichever casing we try.
+        # But users can change that behavior with configuration
+        # (ref https://learn.microsoft.com/en-us/windows/wsl/case-sensitivity),
+        # thus  better prevent it by always relying on case insensitive match.
+        existing_files = {
+            ef.name.casefold(): ef.path for ef in os.scandir(seed_dir)
+        }
+        if not existing_files:
+            LOG.error("%s directory is empty", seed_dir)
+            return None
+
+        folded_names = [
+            f.casefold()
+            for f in candidate_user_data_file_names(self.instance_name)
+        ]
+        for filename in folded_names:
+            if filename in existing_files.keys():
+                return PurePath(existing_files[filename])
+
+        LOG.error(
+            "%s doesn't contain any of the expected user-data files", seed_dir
+        )
+        return None
+
+    def _get_data(self) -> bool:
+        self.vendordata_raw = None
+
+        self.metadata = dict()
+        m_id = machine_id()
```

Comment:
So, if someone removes or clears /etc/machine-id  or sets it to `uninitialized` triggering a recreation of /etc/machine-id on next boot, cloud-init will re-run and re-apply all user-data discovered on this machine, regenerating SSH host keys etc.   I think this is reasonable, but I wanted to call it out in case we have concerns about a /etc/machine-id manipulation forcing cloud-init to re-run on next boot. @holmanb @TheRealFalcon any concerns about this?

Reply:
> if someone removes or clears /etc/machine-id or sets it to uninitialized triggering a recreation of /etc/machine-id on next boot, cloud-init will re-run and re-apply all user-data discovered on this machine, regenerating SSH host keys etc. 

Machine-id changing is not a random event. It is an intentional act by an image builder or admin to say "treat this like a new instance boot - do the first boot stuffs". I think that this is exactly what we want cloud-init to do when machine-id changes - regardless of datasource.

From [man machine-id](https://www.freedesktop.org/software/systemd/man/latest/machine-id.html#)

> For operating system images which are created once and used on multiple machines, for example for containers or in the cloud, /etc/machine-id should be either missing or an empty file in the generic file system image (the difference between the two options is described under "First Boot Semantics" below). An ID will be generated during boot and saved to this file if possible

Getting instance id from the network is [known to be problematic and flaky](https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/1885527) (when were networks ever reliable?). A local file would be more reliable than instance-id.

I'm not saying we should replace instance-id with machine-id (at least right now), but I am saying that I think that if machine-id has changed, that cloud-init should probably treat it as first boot - regardless of whether or not the cloud thinks the instance id has changed.

Reply:
This also raises a point about init systems - `/etc/machine-id` isn't going to be created by Alpine, for example, since it doesn't use systemd. This should really be implemented in the base distro class in `cloudinit/distros/__init__.py` so that other distros can override behavior in a way that makes sense to them.

Reply:
@holmanb

> This also raises a point about init systems - /etc/machine-id isn't going to be created by Alpine, for example, since it doesn't use systemd.

It will be created on Alpine **if** the machine-id init.d service is enabled and runs before the DataSource runs: https://git.alpinelinux.org/aports/tree/main/openrc/machine-id.initd